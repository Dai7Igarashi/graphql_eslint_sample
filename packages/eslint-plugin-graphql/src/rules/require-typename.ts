import { GraphQLESLintRule } from '@graphql-eslint/eslint-plugin';
import type { GraphQLESTreeNode } from '@graphql-eslint/eslint-plugin/esm/estree-converter/types.d.mts';
import { Kind, SelectionNode } from 'graphql';

export const RULE_ID = 'require-typename';

export const rule: GraphQLESLintRule = {
  meta: {
    type: 'problem',
    fixable: 'code',
    schema: [],
    hasSuggestions: true,
    docs: {
      category: 'Operations',
      description:
        'Require __typename in all fragment and inline-fragment field',
      url: '',
      examples: [
        {
          title: 'Incorrect (inline-fragment)',
          code: /* GraphQL */ `
            query getEntry($amebaId: String!, $entryId: Int64!) {
              entry(amebaId: $amebaId, entryId: $entryId) {
                ... on Entry {
                  # __typename is required in inline-fragment
                  title
                }
                ... on Failure {
                  # __typename is required in inline-fragment
                  code
                }
              }
            }

            fragment EntryFields on Entry {
              __typename
              title
            }
          `,
        },
        {
          title: 'Incorrect (fragment)',
          code: /* GraphQL */ `
            query getEntry($amebaId: String!, $entryId: Int64!) {
              entry(amebaId: $amebaId, entryId: $entryId) {
                ... on Entry {
                  ...EntryFields
                }
                ... on Failure {
                  # __typename is required in inline-fragment
                  code
                }
              }
            }

            fragment EntryFields on Entry {
              # __typename is required in fragment.
              title
            }
          `,
        },
        {
          title: 'Correct(inline-fragment)',
          code: /* GraphQL */ `
            query getEntry($amebaId: String!, $entryId: Int64!) {
              entry(amebaId: $amebaId, entryId: $entryId) {
                ... on Entry {
                  __typename
                  title
                }
                ... on Failure {
                  __typename
                  code
                }
              }
            }
          `,
        },
        {
          title: 'Correct (fragment)',
          code: /* GraphQL */ `
            query getEntry($amebaId: String!, $entryId: Int64!) {
              entry(amebaId: $amebaId, entryId: $entryId) {
                ... on Entry {
                  ...EntryFields
                }
                ... on Failure {
                  __typename
                  code
                }
              }
            }

            fragment EntryFields on Entry {
              __typename
              title
            }
          `,
        },
      ],
    },
  },
  create(context) {
    function hasTypenameInEachField(
      node: GraphQLESTreeNode<SelectionNode>,
      values: string[],
    ): void {
      const hasTypename = values.includes('__typename');
      if (!hasTypename) {
        context.report({
          node,
          message: 'Require __typename in all fragment and inline-fragment',
          suggest: [
            {
              desc: 'Add __typename before field',
              fix(fixer) {
                const necessaryWhitespace = ' '.repeat(node.loc.start.column);
                return fixer.insertTextBeforeRange(
                  [node.range[0], node.range[1]],
                  `__typename\n${necessaryWhitespace}`,
                );
              },
            },
          ],
        });
      }
    }

    return {
      /** inline-fragmentに対するlint */
      SelectionSet: function (node) {
        for (const selection of node.selections) {
          if (selection.kind === Kind.INLINE_FRAGMENT) {
            const values: string[] = [];
            const inlineFragmentSelections = selection.selectionSet.selections;
            for (const inlineFragmentSelection of inlineFragmentSelections) {
              if (inlineFragmentSelection.kind === Kind.FIELD) {
                values.push(inlineFragmentSelection.name.value || '');
              }
            }
            hasTypenameInEachField(inlineFragmentSelections[0], values);
          }
        }
      },
      /** fragmentに対するlint */
      FragmentDefinition: function (node) {
        const values: string[] = [];
        for (const selection of node.selectionSet.selections) {
          if (selection.kind === Kind.FIELD) {
            values.push(selection.name.value || '');
          }
        }
        hasTypenameInEachField(node.selectionSet.selections[0], values);
      },
    };
  },
};
