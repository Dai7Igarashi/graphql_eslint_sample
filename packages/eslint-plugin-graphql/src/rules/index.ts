import { rule as requireTypename } from './require-typename';

export const rules = {
  'require-typename': requireTypename,
};
