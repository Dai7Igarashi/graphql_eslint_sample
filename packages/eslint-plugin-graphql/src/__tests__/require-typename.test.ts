import { GraphQLRuleTester } from '@graphql-eslint/eslint-plugin';
import { rule, RULE_ID } from '../rules/require-typename';

const ruleTester = new GraphQLRuleTester();

ruleTester.runGraphQLTests(RULE_ID, rule, {
  valid: [
    {
      name: '__typename in inline-fragment',
      code: /* GraphQL */ `
        query getEntry($amebaId: String!, $entryId: Int64!) {
          entry(amebaId: $amebaId, entryId: $entryId) {
            ... on Entry {
              __typename
              title
            }
            ... on Failure {
              __typename
              code
            }
          }
        }
      `,
    },
    {
      name: '__typename in fragment',
      code: /* GraphQL */ `
        query getEntry($amebaId: String!, $entryId: Int64!) {
          entry(amebaId: $amebaId, entryId: $entryId) {
            ...EntryFields
            ... on Failure {
              __typename
              code
            }
          }
        }

        fragment EntryFields on Entry {
          __typename
          title
        }
      `,
    },
    {
      name: '__typename except inline-fragment or fragment',
      code: /* GraphQL */ `
        query getEntry($amebaId: String!, $entryId: Int64!) {
          entry(amebaId: $amebaId, entryId: $entryId) {
            id
          }
        }
      `,
    },
  ],
  invalid: [
    {
      name: 'no __typename in inline-fragment',
      code: /* GraphQL */ `
        query getEntry($amebaId: String!, $entryId: Int64!) {
          entry(amebaId: $amebaId, entryId: $entryId) {
            ... on Entry {
              title
            }
            ... on Failure {
              code
            }
          }
        }
      `,
      errors: [
        { message: 'Require __typename in all fragment and inline-fragment' },
        { message: 'Require __typename in all fragment and inline-fragment' },
      ],
    },
  ],
});
