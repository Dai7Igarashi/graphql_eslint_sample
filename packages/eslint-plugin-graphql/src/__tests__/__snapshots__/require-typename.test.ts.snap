// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`no __typename in inline-fragment 1`] = `
"#### ⌨️ Code

       1 |         query getEntry($amebaId: String!, $entryId: Int64!) {
       2 |           entry(amebaId: $amebaId, entryId: $entryId) {
       3 |             ... on Entry {
       4 |               title
       5 |             }
       6 |             ... on Failure {
       7 |               code
       8 |             }
       9 |           }
      10 |         }

#### ❌ Error 1/2

      3 |             ... on Entry {
    > 4 |               title
        |               ^^^^^ Require __typename in all fragment and inline-fragment
      5 |             }

#### 💡 Suggestion: Add __typename before field

     1 |         query getEntry($amebaId: String!, $entryId: Int64!) {
     2 |           entry(amebaId: $amebaId, entryId: $entryId) {
     3 |             ... on Entry {
     4 |               __typename
     5 |               title
     6 |             }
     7 |             ... on Failure {
     8 |               code
     9 |             }
    10 |           }
    11 |         }

#### ❌ Error 2/2

      6 |             ... on Failure {
    > 7 |               code
        |               ^^^^ Require __typename in all fragment and inline-fragment
      8 |             }

#### 💡 Suggestion: Add __typename before field

     1 |         query getEntry($amebaId: String!, $entryId: Int64!) {
     2 |           entry(amebaId: $amebaId, entryId: $entryId) {
     3 |             ... on Entry {
     4 |               title
     5 |             }
     6 |             ... on Failure {
     7 |               __typename
     8 |               code
     9 |             }
    10 |           }
    11 |         }"
`;
